hazelcast-sql/src/test/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastIntegerTypeTest.java

```diff
@@ -0,0 +1,283 @@
+                assertType(INTEGER, i, false, type);
+                assertType(INTEGER, i, true, nullableType);
+
+                assertFalse(HazelcastIntegerType.canOverflow(type));
+                assertFalse(HazelcastIntegerType.canOverflow(nullableType));
+
+                assertEquals(i, HazelcastIntegerType.noOverflowBitWidthOf(type));
+                assertEquals(i, HazelcastIntegerType.noOverflowBitWidthOf(nullableType));
+            } else if (i < Long.SIZE) {
+                assertType(BIGINT, i, false, type);
+                assertType(BIGINT, i, true, nullableType);
+
+                assertFalse(HazelcastIntegerType.canOverflow(type));
+                assertFalse(HazelcastIntegerType.canOverflow(nullableType));
+
+                assertEquals(i, HazelcastIntegerType.noOverflowBitWidthOf(type));
+                assertEquals(i, HazelcastIntegerType.noOverflowBitWidthOf(nullableType));
+            } else {
+                assertType(BIGINT, Long.SIZE, false, type);
+                assertType(BIGINT, Long.SIZE, true, nullableType);
+
+                assertTrue(HazelcastIntegerType.canOverflow(type));
+                assertTrue(HazelcastIntegerType.canOverflow(nullableType));
+
+                assertEquals(Long.SIZE - 1, HazelcastIntegerType.noOverflowBitWidthOf(type));
+                assertEquals(Long.SIZE - 1, HazelcastIntegerType.noOverflowBitWidthOf(nullableType));
+            }
+        }
+    }
+
+    @Test
+    public void testSupports() {
+        for (SqlTypeName typeName : SqlTypeName.values()) {
+            assertEquals(typeName == TINYINT || typeName == SMALLINT || typeName == INTEGER || typeName == BIGINT,
+                    HazelcastIntegerType.supports(typeName));
+        }
+    }
+
+    @Test
+    public void testBitWidthOfLong() {
+        assertEquals(0, HazelcastIntegerType.bitWidthOf(0));
+        assertEquals(1, HazelcastIntegerType.bitWidthOf(1));
+        assertEquals(1, HazelcastIntegerType.bitWidthOf(-1));
+        assertEquals(2, HazelcastIntegerType.bitWidthOf(2));
+        assertEquals(2, HazelcastIntegerType.bitWidthOf(-2));
+        assertEquals(10, HazelcastIntegerType.bitWidthOf(555));
+        assertEquals(10, HazelcastIntegerType.bitWidthOf(-555));
+
+        assertEquals(Long.SIZE - 1, HazelcastIntegerType.bitWidthOf(Long.MAX_VALUE));
+        assertEquals(Long.SIZE - 1, HazelcastIntegerType.bitWidthOf(Long.MIN_VALUE));
+
+        assertEquals(Integer.SIZE - 1, HazelcastIntegerType.bitWidthOf(Integer.MAX_VALUE));
+        assertEquals(Integer.SIZE - 1, HazelcastIntegerType.bitWidthOf(Integer.MIN_VALUE));
+
+        assertEquals(Short.SIZE - 1, HazelcastIntegerType.bitWidthOf(Short.MAX_VALUE));
+        assertEquals(Short.SIZE - 1, HazelcastIntegerType.bitWidthOf(Short.MIN_VALUE));
+
+        assertEquals(Byte.SIZE - 1, HazelcastIntegerType.bitWidthOf(Byte.MAX_VALUE));
+        assertEquals(Byte.SIZE - 1, HazelcastIntegerType.bitWidthOf(Byte.MIN_VALUE));
```

Comment:
If I understand [this comment](https://github.com/taburet/hazelcast/blob/81c4f6c7c99845ed6246413cb1e14f7c51b7a9e5/hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastIntegerType.java#L42-L43) correctly, `bitWidthOf(N) = bitWidthOf(-N)`. But it doesn't hold here, `bitWidthOf(128) = 8` and `bitWidthOf(-128) = 7`. Is this correct?

Reply:
Maybe [this](https://github.com/taburet/hazelcast/blob/81c4f6c7c99845ed6246413cb1e14f7c51b7a9e5/hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastReturnTypes.java#L123-L127) is related

Reply:
Yes, this is correct, documented in 1f0b562628e25e8cd2941ff956afedd5fd8c5d42